src/Build/BackEnd/Components/Logging/EventSourceSink.cs

```diff
@@ -527,6 +544,48 @@ private void RaiseBuildFinishedEvent(object sender, BuildFinishedEventArgs build
             RaiseStatusEvent(sender, buildEvent);
         }
 
+        /// <summary>
+        /// Raises a "build canceled" event to all registered loggers.
+        /// </summary>
+        /// <param name="sender">sender of the event</param>
+        /// <param name="buildEvent">BuildCanceledEventArgs</param>
+        /// <exception cref="LoggerException">When EventHandler raises an logger exception the LoggerException is rethrown</exception>
+        /// <exception cref="InternalLoggerException">Any exceptions which are not LoggerExceptions are wrapped in an InternalLoggerException</exception>
+        /// <exception cref="Exception">ExceptionHandling.IsCriticalException exceptions will not be wrapped</exception>
+        private void RaiseBuildCanceledEvent(object sender, BuildCanceledEventArgs buildEvent)
+        {
+            if (BuildCanceled != null)
+            {
+                try
+                {
+                    BuildCanceled(sender, buildEvent);
+                }
+                catch (LoggerException)
+                {
+                    // if a logger has failed politely, abort immediately
+                    // first unregister all loggers, since other loggers may receive remaining events in unexpected orderings
+                    // if a fellow logger is throwing in an event handler.
+                    this.UnregisterAllEventHandlers();
+                    throw;
+                }
+                catch (Exception exception)
+                {
+                    // first unregister all loggers, since other loggers may receive remaining events in unexpected orderings
+                    // if a fellow logger is throwing in an event handler.
+                    this.UnregisterAllEventHandlers();
+
+                    if (ExceptionHandling.IsCriticalException(exception))
+                    {
+                        throw;
+                    }
+
+                    InternalLoggerException.Throw(exception, buildEvent, "FatalErrorWhileLogging", false);
+                }
+            }
+
+            RaiseStatusEvent(sender, buildEvent);
+        }
```

Comment:
Oh - this code repetition is unfortunate. Not a problem of this PR - but we should remove it eventualy.

Reply:
Not only this part, but I wanted this PR to be focused on fixing the issue.

Reply:
For the Canceled event, what is the benefit of this special-case handling over letting it fall through to `RaiseBuildStatusEvent`?

Reply:
I modified the PR to use the simpler path and treat this event as a general status event.